nixos/modules/services/networking/v2ray.nix

```diff
@@ -56,35 +61,49 @@ with lib;
 
   };
 
-  config = let
-    cfg = config.services.v2ray;
-    configFile = if cfg.configFile != null
-      then cfg.configFile
-      else pkgs.writeTextFile {
-        name = "v2ray.json";
-        text = builtins.toJSON cfg.config;
-        checkPhase = ''
-          ${cfg.package}/bin/v2ray test -c $out
-        '';
+  config =
+    let
+      cfg = config.services.v2ray;
+      configFile =
+        if cfg.configFile != null then
+          cfg.configFile
+        else
+          pkgs.writeTextFile {
+            name = "v2ray.json";
+            text = builtins.toJSON cfg.config;
+            checkPhase = ''
+              ${cfg.package}/bin/v2ray test -c $out
+            '';
+          };
+
+    in
+    mkIf cfg.enable {
+      assertions = [
+        {
+          assertion = (cfg.configFile == null) != (cfg.config == null);
+          message = "Either but not both `configFile` and `config` should be specified for v2ray.";
+        }
+      ];
+
+      environment.etc."v2ray/config.json".source = configFile;
+
+      systemd.packages = [ cfg.package ];
+
+      systemd.services.v2ray = {
+        restartTriggers = [ config.environment.etc."v2ray/config.json".source ];
```

Comment:
Not sure if it's doing anything if the config is specified as a path: `services.v2ray.configFile = "/path/to/config.json"`